---
title: "Pertussis and the CMI-PB project"
subtitle: "Educational material for the CMI-PB project"
author:
  - name: Barry J. Grant
    orcid: 0000-0002-2215-4196
    email: bjgrant@ucsd.edu
    affiliations:
      - ref: ucsd      
  - name: Pramod Shinde
    email: pshinde@lji.org
    corresponding: false
    affiliations:
      - ref: lji
  - name: Ferran Soldevila
    email: fsoldevila@lji.org
    corresponding: false
    affiliations:
      - ref: lji
  - name: Bjoern Peters
    orcid: 0000-0002-8457-6693
    email: bpeters@lji.org
    corresponding: true
    affiliations:
      - ref: ucsd      
      - ref: lji
affiliations:
  - id: ucsd
    department: Division of Biological Sciences
    name: University of California, San Diego
    city: La Jolla
    state: CA
  - id: lji
    department: Division of Vaccine Discovery
    name: La Jolla Institute for Immunology
    city: La Jolla
    state: CA
format: 
  html:
    toc: true
    toc-title: Sections
    css: labsheet.css
    include-after-body: js.html
bibliography: references.bib
---

```{r setup, include = FALSE}
library(labsheet)
library(tidyverse)
```

## Background  

Pertussis (more commonly known as **whooping cough**) is a highly contagious respiratory disease caused by the bacterium *Bordetella pertussis* (@fig-pretussis). People of all ages can be infected leading to violent coughing fits followed by a characteristic high-pitched "whoop" like intake of breath. Children have the highest risk for severe complications and death. Recent estimates from the WHO indicate that ~16 million cases and 200,000 infant deaths are due to pertussis annually [@black2010].

![*Bordetella pertussis* attacks cells lining the airways. The rope-like structures shown are cilia, that typically sweep away inhaled dirt and foreign objects. In a pertussis infection, the bacteria use adhesive proteins to stick whilst releasing toxins that damage cells, trigger inflammation and increase mucus production leading to uncontrollable violent coughing.](images/pretussis_in_lungs.png "Bordetella pertussis produces a number of adhesins and toxins, which are important in pathogenesis. Following inhalation, *B. pertussis* adheres to ciliated epithelium in the trachea and bronchi and release these extracellular toxins. "){#fig-pretussis}


> **Optional:** Link to a [video of a child with pertussis](https://www.youtube.com/watch?v=S3oZrMGDMMw) (Note that some viewers may find this upsetting).



## 1. Investigating pertussis cases by year

The United States *Centers for Disease Control and Prevention* (CDC) has been compiling reported pertussis case numbers since 1922 in their *National Notifiable Diseases Surveillance System* (NNDSS). We can view this data on the CDC website here: https://www.cdc.gov/pertussis/surv-reporting/cases-by-year.html


```{r cdcdata, echo=FALSE}
cdc <- data.frame(
                          Year = c(1922L,
                                   1923L,1924L,1925L,1926L,1927L,1928L,
                                   1929L,1930L,1931L,1932L,1933L,1934L,1935L,
                                   1936L,1937L,1938L,1939L,1940L,1941L,
                                   1942L,1943L,1944L,1945L,1946L,1947L,1948L,
                                   1949L,1950L,1951L,1952L,1953L,1954L,
                                   1955L,1956L,1957L,1958L,1959L,1960L,
                                   1961L,1962L,1963L,1964L,1965L,1966L,1967L,
                                   1968L,1969L,1970L,1971L,1972L,1973L,
                                   1974L,1975L,1976L,1977L,1978L,1979L,1980L,
                                   1981L,1982L,1983L,1984L,1985L,1986L,
                                   1987L,1988L,1989L,1990L,1991L,1992L,1993L,
                                   1994L,1995L,1996L,1997L,1998L,1999L,
                                   2000L,2001L,2002L,2003L,2004L,2005L,
                                   2006L,2007L,2008L,2009L,2010L,2011L,2012L,
                                   2013L,2014L,2015L,2016L,2017L,2018L,
                                   2019L),
  Cases = c(107473,
                                   164191,165418,152003,202210,181411,
                                   161799,197371,166914,172559,215343,179135,
                                   265269,180518,147237,214652,227319,103188,
                                   183866,222202,191383,191890,109873,
                                   133792,109860,156517,74715,69479,120718,
                                   68687,45030,37129,60886,62786,31732,28295,
                                   32148,40005,14809,11468,17749,17135,
                                   13005,6799,7717,9718,4810,3285,4249,
                                   3036,3287,1759,2402,1738,1010,2177,2063,
                                   1623,1730,1248,1895,2463,2276,3589,
                                   4195,2823,3450,4157,4570,2719,4083,6586,
                                   4617,5137,7796,6564,7405,7298,7867,
                                   7580,9771,11647,25827,25616,15632,10454,
                                   13278,16858,27550,18719,48277,28639,
                                   32971,20762,17972,18975,15609,18617)
)
```

`r question_start()`
- **Q1.**  With the help of the R "addin" package [**datapasta**](https://milesmcbain.github.io/datapasta/) assign the CDC pertussis case number data to a data frame called `cdc` and use **ggplot** to make a plot of cases numbers over time.

::: {.callout-tip collapse="true"} 
### Hint

To help "scrape" this data from the CDC website you can install the [**datapasta package**](https://milesmcbain.github.io/datapasta/), which will appear under the RStudio "**Addins**" menu button and allow you to web-scrape by copying tables of data from any website or PDF and then in RStudio selecting *"Paste as data.frame"* from the RStudio "**Addins**" menu. 


```{r dunnmycdcplot, eval=FALSE}
ggplot(cdc) +
  aes(___, ___) +
  geom_point() +
  geom____() +
  labs(___)
```



```{r cdcplot, echo=FALSE}
library(ggplot2)

base <- ggplot(cdc) +
  aes(Year, Cases) +
  geom_point() +
  geom_line() +
  labs(title="Pertussis Cases by Year (1922-2019)", y="Number of cases") +
  scale_y_continuous(labels = scales::label_comma()) 

print(base)
```

:::

`r question_end()`

> **Key point:** Pertussis vaccination is, in general, highly effective at preventing the disease. In the pre-vaccine era (before 1946) pertussis was a much more common disease and a major cause of infant mortality [2](https://www.cdc.gov/vaccines/pubs/surv-manual/chpt10-pertussis.html). As we see clearly from analysis of the CDC tracking data above, introduction of the first pertussis vaccination in the United States in 1946 resulted in a dramatic reduction in the number of yearly cases from > 200,000 in the 1940s to < 2,000 in the 1970s. 


## 2. A tale of two vaccines (wP & aP)

Two types of pertussis vaccines have been developed: **whole-cell pertussis (wP)** and **acellular pertussis (aP)**. The first vaccines were composed of ‘whole cell’ (wP) inactivated bacteria. The latter aP vaccines use purified antigens of the bacteria (the most important pertussis components for our immune system, see @fig-vacine). These aP vaccines were developed to have less side effects than the older wP vaccines and are now the only form administered in the United States.  


![Whole-cell pertussis vaccines are inactivated bacteria with many antigens, including endotoxins, which trigger broad immunity. The acellular vaccines use only a few antigens, including a weakened pertussis toxin, to limit side effects.](images/vacine.png "Whole-cell pertussis vaccines are inactivated bacteria with many antigens, including endotoxins, which trigger broad immunity. The acellular vaccines use only a few antigens, including a weakened pertussis toxin, to limit side effects."){#fig-vacine width="300"}


> *Side-note*: Currently in United States aP vaccines are administered to infants in combination with those for **d**iphtheria (D) and **t**etanus (T) in the so-called **DTaP vaccine** combination.  A distinct **Tdap** formulation, with reduced doses of diphtheria and pertussis, is administered as a boost for preteens, teens, and adults.

For motivated readers there is a nice historical account of the wP to aP vaccine switch in the US in Klein (2014) [@klein2014]


Let's return to our CDC data plot and examine what happened after the switch to the acellular pertussis (aP) vaccination program.




`r question_start()`
- **Q2.** Using the ggplot `geom_vline()` function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?  


::: {.callout-tip collapse="true"} 
### Hint

Your plot should look something like the one below where we take our plot from Q1  and minimally add two `geom_vline(xintercept=___)` function calls. Here I also add optional text labels with `geom_text()` to help annotate the lines.

```{r echo=FALSE}
base + geom_vline(xintercept=1946, col="blue", linetype = 2) +
  geom_vline(xintercept=c(1996), col="red", linetype = 2) +
  geom_text(mapping = aes(x = 1946,
                          y = max(cdc),
                          label = "wP",
                          hjust = -0.1,
                          vjust = 0), color="blue", size=3.4  ) +
   geom_text(mapping = aes(x = 1996,
                          y = max(cdc),
                          label = "aP",
                          hjust = -0.1,
                          vjust = 0), color="red", size=3.4 ) 
```

:::


- **Q3.** Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

::: {.callout-tip collapse="true"} 
### Hint

It is clear from the CDC data that pertussis cases are once again increasing. For example, we can see that in 2012 the CDC reported 48,277 cases of pertussis in the United States. This is the largest number of cases reported since 1955, when 62,786 cases were reported. The pertussis field has several hypotheses for the resurgence of pertussis including (in no particular order): **1)** more sensitive PCR-based testing, **2)** vaccination hesitancy **3)** bacterial evolution (escape from vaccine immunity), **4)** waning of immunity in adolescents originally primed as infants with the newer aP vaccine as compared to the older wP vaccine.


:::

`r question_end()`
 
> **Key-point:** Despite high levels of acellular pertussis (aP) vaccination, the United States and other countries are now experiencing a significant resurgence in pertussis cases with large outbreaks now once again a major public health concern.

<span style="color: green;">
**Additional points for discussion:** How are vaccines currently approved?
</span>

<span style="color: green;"><ul> 
<li> Typically we first examine *‘Correlates of protection’* which are things that can be measured within weeks or months after vaccination, and which are thought to correlate with increased protection from disease. For the aP vaccine this was  an induction of antibodies against pertussis toxin (PT) in infants at equivalent levels to those induced by the wP vaccine. The aP vaccines also had less side effects (reduction of sore arms, fever and pain).</li>  
<li> Testing for protection induced by a new vaccine requires a lot of people exposed to the pathogen (like in a pandemic).</li>  
<li> It is impossible to discover a effect 10 years post vaccination in the current trial system.</li>
<li> It is unclear what differentiates people that have been primed with aP vs. wP long term.</li>
<li> The **CMI-PB project** is an attempt to make data on this question open and examinable by all.</li>
</ul></span>

## 3. Exploring CMI-PB data

**Why is this vaccine-preventable disease on the upswing?** To answer this question we need to investigate the mechanisms underlying waning protection against pertussis. This requires evaluation of pertussis-specific immune responses over time in wP and aP vaccinated individuals. 

The new and ongoing [**CMI-PB project**](https://www.cmi-pb.org/) aims to provide the scientific community with this very information. In particular, CMI-PB tracks and makes freely available long-term humoral and cellular immune response data for a large number of individuals who received either DTwP or DTaP combination vaccines in infancy followed by Tdap booster vaccinations. This includes complete API access to longitudinal RNA-Seq, AB Titer, Olink, and live cell assay results directly from their website: https://www.cmi-pb.org/ 

> **Key-point**: For the CMI-PB project a booster shot is a proxy for what happens when you get infected.	



### The CMI-PB API returns JSON data

The CMI-PB API (like most APIs) sends responses in JSON format. Briefly, JSON data is formatted as a series of **key-value pairs**, where a particular word ("key") is associated with a particular value. An example of the JSON format for Ab titer data is shown below:

```
{
"specimen_id":1,
 "isotype":"IgG",
 "is_antigen_specific":true,
 "antigen":"PT",
 "ab_titer":68.5661390514946,
 "unit":"IU/ML",
 "lower_limit_of_detection":0.53
 } 

```


To read these types of files into R we will use the `read_json()` function from the **jsonlite** package. Note that if you want to do more advanced querys of APIs directly from R you will likely want to explore the more full featured  **rjson** package. The big advantage of using jsonlite for our current purposes is that it can simplify JSON key-value pair arrays into R data frames without much additional effort on our part.

```{r message=FALSE}
# Allows us to read, write and process JSON data
library(jsonlite)

```

Let's now read the main **subject** database table from the CMI-PB API. You can find out more about the content and format of this and other tables here: https://www.cmi-pb.org/blog/understand-data/


```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```


> **Key-point:** The `subject` table provides metadata about each individual in the study group. For example, their infancy vaccination type, biological sex, year of birth, time of boost etc.


```{r}
head(subject, 3)
```


`r question_start()`
- **Q4.** How many aP and wP infancy vaccinated subjects are in the dataset?

::: {.callout-tip collapse="true"}
### Hint

For these types of questions we can use our old friend the **table()** function, which will *tabulate* the number of unique entries for us:  

```{r}
table(subject$infancy_vac)
```
:::

- **Q5.** How many Male and Female subjects/patients are in the dataset?

```{r eval=FALSE, echo=FALSE}
table(subject$biological_sex)
```

- **Q6.** What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc...)?

::: {.callout-tip collapse="true"} 
### Hint

```{r eval=FALSE}
table(subject$___, subject$___) )
```

```{r echo=FALSE}
knitr::kable( table(subject$biological_sex, subject$race) )
```
:::

`r question_end()`


### Side-Note:  Working with dates

Two of the columns of `subject` contain dates in the Year-Month-Day format. Recall from our last mini-project that dates and times can be annoying to work with at the best of times. However, in R we have the excellent lubridate package, which can make life allot easier. Here is a quick example to get you started:

```{r message=FALSE}
library(lubridate)
```

What is today’s date (at the time I am writing this obviously)
```{r}
today()
```

How many days have passed since new year 2000 

```{r}
today() - ymd("2000-01-01")
```

What is this in years?

```{r}
time_length( today() - ymd("2000-01-01"),  "years")
```

Note that here we are using the `ymd()` function to tell lubridate the format of our particular date and then the `time_length()` function to convert days to years.

`r question_start()`
- **Q7.** Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different? 

::: {.callout-tip collapse="true"} 
### Hint  

Using `ymd()` function on the appropriate column of our `subject` data frame together with today's date we can calculate the age in days for subjects. We can optionally store this back into our `subject` data frame for later use:    
```{r echo=FALSE}
# Use todays date to calculate age in days
subject$age <- today() - ymd(subject$year_of_birth)
```

```{r eval=FALSE}
# Use todays date to calculate age in days
subject$age <- today() - ___
```

Next we can use **dplyr**'s `filter()` function to limit ourselves to a particular subset of subjects to examine the 6 number summary of their age in years: 
```{r}
library(dplyr)

ap <- subject %>% filter(infancy_vac == "aP")

round( summary( time_length( ap$age, "years" ) ) )
```

```{r eval=FALSE}
# wP
wp <- subject %>% ___(___ == "wP")
round( summary( time_length( wp$age, "years" ) ) )

```

```{r echo=FALSE}
# wP
wp <- subject %>% filter(infancy_vac == "wP")
round( summary( time_length( wp$age, "years" ) ) )
```
:::




- **Q8.** Determine the age of all individuals at time of boost?

::: {.callout-tip collapse="true"} 
### Hint

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```

:::




- **Q9.** With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different? 

```{r message=FALSE}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) +
  xlab("Age in years")
```

::: {.callout-tip collapse="true"} 
### Hint

These groups are clearly different. It is not necessary but we could calculate a p-value...
```{r}
# Or use wilcox.test() 
x <- t.test(time_length( wp$age, "years" ),
       time_length( ap$age, "years" ))

x$p.value
```
:::

`r question_end()`


### Joining multiple tables

Read the specimen and ab_titer tables into R and store the data as `specimen` and `titer` named data frames.

```{r echo=FALSE}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/plasma_ab_titer", simplifyVector = TRUE) 
```

```{r eval=FALSE}
# Complete the API URLs...
specimen <- read_json("___/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/___", simplifyVector = TRUE) 
```



To know whether a given `specimen_id` comes from an aP or wP individual we need to link (a.k.a. "join" or merge) our `specimen` and `subject` data frames. The excellent **dplyr** package (that we have used previously) has a family of `join()` functions that can help us with this common task: 

`r question_start()`
- **Q9.** Complete the code to join `specimen` and `subject` tables to make a new merged data frame containing all specimen records along with their associated subject details: 

```{r eval=FALSE}
meta <- ____join(___, subject)
dim(meta)
head(meta)
```

```{r echo=FALSE, message=FALSE}
meta <- inner_join(specimen, subject)
dim(meta)
```

::: {.callout-tip collapse="true"} 
### Hint

To answer this question you need to chose the correct `join()` function (from `inner_join()`, `left_join()`, `right_join()` and `full_join()` options) as well as select the order of `x` and `y` input arguments.

::: {.content-visible unless-format="pdf"}

![](images/inner-join.gif)


![](images/full-join.gif)


:::

For a nice description of different join types see: https://www.garrickadenbuie.com/project/tidyexplain/ from where the animated images in the HTML version of this document were taken.

:::

- **Q10.** Now using the same procedure join `meta` with `titer` data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r eval=FALSE}
abdata <- inner_join(___, meta)
dim(abdata)
```

```{r echo=FALSE}
abdata <- inner_join(titer, meta)
dim(abdata)
```

- **Q11.** How many specimens (i.e. entries in `abdata`) do we have for each `isotype`? 

::: {.callout-tip collapse="true"} 
### Hint

We can use our old friend the `table()` function on the `isotype` column to answer this question: 

```{r}
table(abdata$isotype)
```
:::

- **Q12.** What are the different `$dataset` values in `abdata` and what do you notice about the number of rows for the most "recent" dataset? 

::: {.callout-tip collapse="true"} 
### Hint

You can again use `table()` this time on the the `dataset` column to help answer this question.

:::

`r question_end()`


# 4. Examine IgG Ab titer levels

Now using our joined/merged/linked `abdata` dataset `filter()` for IgG `isotype`.  

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

`r question_start()`
- **Q13.** Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens: 

```{r eval=FALSE}
ggplot(igg) +
  aes(___, antigen) +
  geom____() + 
    xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)
```


```{r echo=FALSE, warning=FALSE}
ggplot(igg) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() + 
  xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)

```
   

- **Q14.** What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others? 

::: {.callout-tip collapse="true"} 
### Hint

You can use the CMI-PB website search functionality and *Terminology Browser* (under development) to find out about each antigen. Note that this is still work in progress. 
:::

`r question_end()`

We can attempt to examine differences between wP and aP here by setting color and/or facet values of the plot to include `infancy_vac` status (see below). However these plots tend to be rather busy and thus hard to interpret easily. 

```{r warning=FALSE}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()

```

Another version of this plot adding `infancy_vac` to the faceting:

```{r}
igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)

```

> Side-note: If you don't like the overlapping x axis labels (and who would?) you can add a `theme()` layer where you set the text angle and horisontal adjustment relative to the axis. For example: `theme(axis.text.x = element_text(angle = 45, hjust=1))`


`r question_start()`
- **Q15.** Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a "control" antigen (**"OVA"**, that is not in our vaccines) and a clear antigen of interest (**"PT"**, Pertussis Toxin, one of the key virulence factors produced by the bacterium *B. pertussis*).  


```{r eval=FALSE}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI____, col=infancy_vac) +
  geom_boxplot(show.legend = ___) +
  facet_wrap(vars(visit)) +
  theme_bw()
```


```{r echo=FALSE}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  labs(title="OVA antigen levels per visit (aP red, wP teal)") +
  theme_bw() +
  theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) 

```
  
and the same for `antigen=="FIM2/3"` 

```{r eval=FALSE}
filter(igg, antigen==___) %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = ___) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r echo=FALSE}
filter(igg, antigen=="PT") %>% 
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  labs(title="PT antigen levels per visit (aP red, wP teal)") +
  theme_bw() +
  theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) 

```
  

- **Q16.** What do you notice about these two antigens time courses and the PT data in particular?

::: {.callout-tip collapse="true"} 
### Hint

PT levels clearly rise over time and far exceed those of OVA. They also appear to peak at visit 5 and then decline. This trend appears similar for wP and aP subjects.
:::

- **Q17.** Do you see any clear difference in aP vs. wP responses? 
`r question_end()`

Lets finish this section by looking at the 2021 dataset IgG PT antigen levels time-course:

```{r}
abdata.21 <- abdata %>% filter(dataset == "2021_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")

```

`r question_start()`
- **Q18.** Does this trend look similar for the 2020 dataset?  
`r question_end()`





# 5. Obtaining CMI-PB RNASeq data

For RNA-Seq data the API query mechanism quickly hits the web browser interface limit for file size. We will present alternative download mechanisms for larger CMI-PB datasets in the next section. However, we can still do "targeted" RNA-Seq querys via the web accessible API. 

For example we can obtain RNA-Seq results for a specific ENSEMBLE gene identifier or multiple identifiers combined with the & character:

```
# For example use the following URL
https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7
```

The link above is for the key gene involved in expressing any IgG1 antibody, namely the [IGHG1 gene](https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000211896;r=14:105736343-105743071). Let's read available RNA-Seq data for this gene into R and investigate the time course of it's gene expression values. 


```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

To facilitate further analysis we need to "join" the `rna` expression data with our metadata `meta`, which is itself a join of sample and specimen data. This will allow us to look at this genes TPM expression values over aP/wP status and at different visits (i.e. times):


```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

`r question_start()`
- **Q19.** Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of `visit` vs. `tpm`).

```{r eval=FALSE}
ggplot(ssrna) +
  aes(___, ___, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

```{r echo=FALSE}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```


- **Q20.**: What do you notice about the expression of this gene (i.e. when is it at it's maximum level)?

- **Q21.** Does this pattern in time match the trend of antibody titer data? If not, why not? 

::: {.callout-tip collapse="true"} 
### Hint

Have a look back at your plots for Q15 (antigen levels) and compare to your expression levels. Cells make antibodies, which are long lived.

:::

`r question_end()`

We can dig deeper and color and/or facet by `infancy_vac` status:

```{r}

ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

There is however no obvious wP vs. aP differences here even if we focus in on a particular visit:

```{r}

ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```
 



# 6. Working with larger datasets [OPTIONAL] 

As API based RNA-Seq querys can quickly become large requests, CMI-PB makes CSV files available for download. We will take their "2020 longitudinal RNA-Seq data" file (named `2020LD_rnaseq.csv`) from here:
https://www.cmi-pb.org/downloads/cmipb_challenge_datasets/1st_cmipb_challenge/04272022/

> **Note:** This file is over 321MB so the download may take some time over slow internet connections. 


```{r}
# Change for your downloaded file path
rnaseq <- read.csv("~/Downloads/2020LD_rnaseq.csv")

head(rnaseq,3)
```

This is rather large, look at the number of rows we have here:
```{r}
dim(rnaseq)
```

### Working with long format data

Note that our `rnadata` is in so-called **"long" format** rather than the more conventional **"wide" format** tables of expression data that we have worked with in the past (where rows represent genes and different columns represent counts in different experiments with a column per experiment). Here we have genes in the rows but have we counts for all experiments collated together in one column. Along with this we have our now familiar `specimen_id` column to tell us which experiment the values come from.  

Lets have a look at how many genes we have reported for each `specimen_id` with our old friend the `table()` function.

```{r}
n_genes <- table(rnaseq$specimen_id)
head( n_genes , 10)
```

How many specimens?
```{r}
length(n_genes)
```

Check if there are the same number of genes for each specimen
```{r}
all(n_genes[1]==n_genes)
```

#### Convert to "wide" format

All looks good at this stage so let's convert to wider format with the `pivot_wider()` function from the **tidyr** package:

```{r wideformat}
library(tidyr)

rna_wide <- rnaseq %>% 
  select(versioned_ensembl_gene_id, specimen_id, tpm) %>%
  pivot_wider(names_from = specimen_id, values_from=tpm)

dim(rna_wide)
```


```{r}
head(rna_wide[,1:7], 3)
```


### Filtering

As usual we have lots of zero count genes that can be removed (i.e. filtered out) before further analysis. Generally we will want to remove genes which are expressed at low levels or show only small changes in expression. 

## Key remaining questions

Once you have got this far we can begin to investigate the relationship between mRNA levels on different days (e.g. visit 1 vs. visit 4 etc.) and start to address the next set of pressing questions: 

> **Key-questions:** 
>
> - Is RNA-Seq expression levels predictive of Ab titers?
> - What differentiates aP vs. wP primed individuals? 
> - What about decades after their first immunization? Do you know? 

 
These are all ares of active research. What is clear is that there are immune responses not captured in antibody titers. We are trying to capture the ‘systems level’ response to antigen encounters (here: vaccination). Something must be different in aP vs. wP primed individuals. That is why CMI-PB make all their data available for the wider community to explore and contribute new analysis methods. We invite you to join us in addressing these important open questions!


## About this document

Here we use the `sessionInfo()` function to report on our R systems setup at the time of document execution. 

```{r}
sessionInfo()
```

